### ISSUE TITLE
Hidden rng bugs with multi-PhysicsGenerator

### ISSUE THREAD

**Issue by Andrewwango** [deepinv/deepinv#1336]
I've found two hidden bugs relating to when you have >1 PhysicsGenerators, that severely affect randomness.

## Preliminaries

With any random generation, the goal for the user is to have a) randomness and b) reproducibility. With >1 PhysicsGenerators, the goal is to produce uncorrelated samples from all of them.

There seem to be 4 workflows that one could use to have randomness and reproducibility:

1. Have **one** shared rng instance (per thread) that is passed to every PhysicsGenerator. Every PhysicsGenerator evolves the same rng so no `step` uses the same random state. IMO this should be standard but requires careful planning.
2. Let every PhysicsGenerator have their own unique `rng=torch.Generator().manual_seed(seed)` where `seed` is unique to that generator and rank. Lots of seeds to keep track of!
3. Never set any generators, and let the randomness go wild. Note: this gives you randomness, but not reproducibility (obvs). ⚠️⚠️
4. "I don't care what rngs my PhysicsGenerators have. I ignore it". Then, **every single time** any `.step` is called, manually seed it with `step(seed=seed)` where `seed` is unique to that training iter and rank. A quick patch when your code is messy. ⚠️

## Bug 1

Workflows (1) and (2) are fine. ⚠️⚠️Workflow (3) does not work in deepinv! Let's say we have two dummy PhysicsGenerators which generate horizontal and vertical sigmas of a blur. Or any two generators that use the same sampling function under the hood.

```python
class DummyBlurSigmaXGenerator(dinv.physics.generator.PhysicsGenerator):
    """e.g. generates the horizontal sigma of a blur kernel"""
    def step(self, batch_size = 1, seed = None, **kwargs):
        self.rng_manual_seed(seed)
        assert batch_size == 1
        return {"x": torch.randint(1, 1000, (1,), generator=self.rng, **self.factory_kwargs)}

class DummyBlurSigmaYGenerator(dinv.physics.generator.PhysicsGenerator):
    """e.g. generates the vertical sigma of an anisotropic blur kernel"""
    def step(self, batch_size = 1, seed = None, **kwargs):
        self.rng_manual_seed(seed)
        assert batch_size == 1
        return {"y": torch.randint(1, 1000, (1,), generator=self.rng, **self.factory_kwargs) / 1000}

gen1 = DummyBlurSigmaXGenerator()
gen2 = DummyBlurSigmaYGenerator()

for _ in range(3):
    print(gen1.step()['x'], gen2.step()['y'])

{'x': tensor([361.])} {'y': tensor([0.3610])}
{'x': tensor([930.])} {'y': tensor([0.9300])}
{'x': tensor([237.])} {'y': tensor([0.2370])}
```
The two generators step in unison, meaning they can't explore the full diversity. Same story for GeneratorMixtures and sums:
```python
gen = dinv.physics.generator.GeneratorMixture([DummyBlurSigmaXGenerator(), DummyBlurSigmaYGenerator()], probs=[0.5, 0.5])
for _ in range(6):
    print(gen.step())

{'x': tensor([930.])} # side bug: why does this skip 361 but no others? 
{'y': tensor([0.9300])}
{'x': tensor([237.])}
{'x': tensor([215.])}
{'y': tensor([0.2370])}
{'x': tensor([568.])}

gen = DummyBlurSigmaXGenerator() + DummyBlurSigmaYGenerator()
for _ in range(3):
    print(gen.step())

{'x': tensor([361.]), 'y': tensor([0.3610])}
{'x': tensor([930.]), 'y': tensor([0.9300])}
{'x': tensor([237.]), 'y': tensor([0.2370])}
```

### Expected behaviour

That in workflow (3) the two generators step independently.

### Proposed fix

The simple reason why this happens is because when you don't pass a rng, we force each PhysicsGenerator to have the default rng which **is seeded identically every time**:

https://github.com/deepinv/deepinv/blob/6cc8c0672da977f1bebb2da3fc822a88aaa8947e/deepinv/physics/generator/base.py#L69-L70

We should either by default give no rng, or a random seed.

## Bug 2: 

⚠️Workflow (4) doesn't work in the case of physics generator sums!

```python
gen = DummyBlurSigmaXGenerator() + DummyBlurSigmaYGenerator()
for i in range(3):
    print(gen.step(seed=42 * i))

{'y': tensor([0.5400]), 'x': tensor([540.])}
{'y': tensor([0.7900]), 'x': tensor([790.])}
{'y': tensor([0.4970]), 'x': tensor([497.])}
```

### Expected behaviour

That in workflow (4) with physics generator sums, the subgenerators evolve uncorrelated.

### Proposed fix

This is caused by the generator sum reseeding every subgenerator identically in step:

https://github.com/deepinv/deepinv/blob/6cc8c0672da977f1bebb2da3fc822a88aaa8947e/deepinv/physics/generator/base.py#L142-L146

We should maybe reseed each PhysicsGenerator differently and non-overlapping:

```python
    def step(self, batch_size: int = 1, seed: int = None, **kwargs):
        result = {}
        for i, gen in enumerate(self.generators):
            for k, v in gen.step(batch_size, seed=None if seed is None else (seed * (i + 1)) % (2 ** 64 - 1), **kwargs).items():
                result[k] = result[k] + v if k in result else v # NOTE: this also the fixes the bug of non-numeric values
        return result
```

## TLDR

The two safe ways today of using multiple PhysicsGenerators randomly and reproducibly:

1. Have **one** shared rng instance (per thread) that is passed to every PhysicsGenerator.
2. Let every PhysicsGenerator have their own unique `rng=torch.Generator().manual_seed(seed)` where `seed` is unique to that generator and rank.

Two ways that **do not work today but should**:

3. Never set any generators, and let the randomness go wild: all PhysicsGenerators step correlatedly.
4. Manually seed all generators every step with `step(seed=seed)`: in the case of summed generators, all PhysicsGenerators step correlatedly.

**Comment by romainvo** [comment]
Nasty errors indeed, thanks @Andrewwango for documenting this! 

I feel like we have two types of problems here: 
1. Being able to control reproducibility efficiently
2. Having true uncorrelated randomness (if not requested)

**1. Controlling reproducibility**
- The shared rng instance is definitely the way to go, having a minimal example snippet implementing this strategy would definitely help users. 
- We discussed this at some point with @jscanvic, the other part of the solution would be to avoid creating any Generator unless the user requests it (this is also what you suggest at the end of "Bug 1" if I understand clearly). The randomness would go wild, as you mentioned, but this also means that a single call to `torch.manual_seed` could effectively control it. This is exactly how randomness is controlled in PyTorch (eg when using `torch.randn`), and I think matching this behavior would improve the workflow's clarity. 

**2. Proper uncorrelated randomness** by default
- Fixing Bug 2 is mandatory as you mentioned, I like your solution :)!
- If we remove the `self.rng` attribute when not specified then it's solved. If we keep a per-generator `self.rng` (even when not requested) then we need to manually seed it (randomly).

To be honest I lean toward the no-generator-by-default approach; it feels simpler to manage.

---

Complementarily (original idea from @jscanvic)

To make randomness control more robust, we could use `torch.random.fork_rng` when passing the `seed` argument in `physics_generator.step`. I see two advantages:
- this is compatible with the "no per-generator rng" solution, i.e. doing `with torch.random.fork_rng(devices=[device]):` forks the rng states of all generators present on `device`, so each stochastic step within this context does not affect the rng states.
- it avoids touching the state of the rng for a single op, i.e. this means the rng can be used across multiple objects and specific control at a time "t" does not influence the rest of the process.

---

A lot of paraphrasing what you already mentioned in your post @Andrewwango, hope this complements what you documented.

**Comment by Andrewwango** [comment]
💯 @romainvo .

Side notes:
- TODO check how this affects multi-NoiseModels
- check how bad the backward compatibility would be with the proposed fixes

**Comment by Andrewwango** [comment]
cc @Zeppo1994 if you're interested, here's a big bug
